gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -3,159 +3,164 @@
+                DiagLazyTensor, select all elements of the omitted batches.
+
+        Returns:
+            An `i`-dim tensor containing the requested elements if all batch index tensors were
+            passed in, otherwise a `bl x ... bk x i` batch of such tensors, where `l = len(batch_indices) + 1`.
+
+        """
+        batch_dims = len(batch_indices)
+        if batch_dims > self._batch_dim:
+            raise RuntimeError(
+                "Received {} batch index tensors, DiagLazyTensor has {} batches".format(batch_dims, self._batch_dim)
+            )
+        full_batch_shape = self._diag.shape[batch_dims:-1]
+        if len(full_batch_shape) == 0:
+            diag_elements = self._diag[batch_indices + (left_indices,)]
         else:
-            res = self._diag.unsqueeze(-1).expand_as(rhs) * rhs
-            return res
+            d = torch.ones_like(left_indices)
+            diag_elements = torch.stack(
+                [
+                    self._diag[batch_indices + tuple(j * d for j in js) + (left_indices,)]
+                    for js in product(*[range(b) for b in full_batch_shape])
+                ]
+            )
+        equal_indices = (left_indices == right_indices).type_as(self._diag)
+        return diag_elements * equal_indices
 
-    def _t_matmul(self, rhs):
-        # Matrix is symmetric
-        return self._matmul(rhs)
+    def _matmul(self, rhs):
+        # to perform matrix multiplication with diagonal matrices we can just
+        # multiply element-wise with the diagonal (using proper broadcasting)
+        if rhs.ndimension() == 1:
+            return self._diag * rhs
+        return self._diag.unsqueeze(-1) * rhs
 
     def _quad_form_derivative(self, left_vecs, right_vecs):
+        # TODO: Use proper batching for input vectors (prepand to shape rathern than append)
         res = left_vecs * right_vecs
         if res.ndimension() > self._diag.ndimension():
             res = res.sum(-1)
         return (res,)
 
     def _size(self):
-        if self._diag.ndimension() == 2:
-            return self._diag.size(0), self._diag.size(-1), self._diag.size(-1)
-        else:
-            return self._diag.size(-1), self._diag.size(-1)
+        return self._diag.shape + self._diag.shape[-1:]
+
+    def _t_matmul(self, rhs):
+        # Diagonal matrices always commute
+        return self._matmul(rhs)
 
     def _transpose_nonbatch(self):
         return self
 
-    def _batch_get_indices(self, batch_indices, left_indices, right_indices):
-        equal_indices = left_indices.eq(right_indices).type_as(self._diag)
-        return self._diag[batch_indices, left_indices] * equal_indices
-
-    def _get_indices(self, left_indices, right_indices):
-        equal_indices = left_indices.eq(right_indices).type_as(self._diag)
-        return self._diag[left_indices] * equal_indices
+    def abs(self):
+        return DiagLazyTensor(self._diag.abs())
```

Comment:
This would also be an interesting method to chuck onto LazyTensors, but I'm not quite sure how yet. Perhaps LazyTensor subclasses should define a method for performing arbitrary element-wise operations. Then the base LazyTensor class can call that subclass method, and say that we want to perform the `abs` operation.

Not for this PR, just thinking out loud :)